File: Java/civitas/src/civitas/MazoSorpresa.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;


public class MazoSorpresa {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    void init(){
        sorpresas = new ArrayList();
        cartasEspeciales = new ArrayList();
        barajada = false;
        usadas = 0;
    }

    public MazoSorpresa() {
        init();
        debug = false;    
    }

    public MazoSorpresa(boolean _debug) {
        this.debug = _debug;
        init();
        if(debug ==true){
            Diario.getInstance().ocurreEvento("Debug en True");
        }
    }
    
    void alMazo(Sorpresa s){
        if(barajada == false){
            sorpresas.add(s);
        }
    }
    
    Sorpresa siguiente() {
        if((barajada == false || usadas == sorpresas.size()) && debug == false){
            barajada = true;
            usadas = 0;
            Collections.shuffle(sorpresas);
        }
        
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(sorpresas.size()-1, ultimaSorpresa);
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        for(int i = 0; i <sorpresas.size(); i++){
            if(sorpresas.get(i) == sorpresa){
                cartasEspeciales.add(sorpresa);
                sorpresas.remove(i);
                Diario.getInstance().ocurreEvento("Carta Sorpresa Inhabilitada For Ever");
            }
        }
    }
    void habilitarCartaEspecial(Sorpresa sorpresa){
        for(int i = 0; i <cartasEspeciales.size(); i++){
            if(cartasEspeciales.get(i) == sorpresa){
                sorpresas.add(sorpresas.size()-1, cartasEspeciales.get(i));
                cartasEspeciales.remove(i);
                Diario.getInstance().ocurreEvento("Carta Sorpresa Habilitada For Ever");
            }
        }
    }

    int size() {
        return sorpresas.size();
    }
    
    Sorpresa getUltimaSorpresa(){
        return this.ultimaSorpresa;
    }
    
    
    Sorpresa getSorpresas(int i){
        return sorpresas.get(i);
    }
    
}

File: Ruby/civitas/lib/mazo_sorpresa.rb
#encoding: utf­8
require_relative 'diario'
require_relative 'sorpresa'
module Civitas
  class MazoSorpresa
    attr_accessor :sorpresas, :ultima_sorpresa
    
    def initialize(_debug)
      @sorpresas = Array.new()
      @barajada = false
      @usadas= 0
      @debug=_debug
      @cartas_especiales=Array.new()
      @ultima_sorpresa
      if(_debug==true)
        Diario.instance.ocurre_evento("Debug activado")
      end 
    end
    
    def self.new_mazo_sorpresa()
      new(false)
    end
    
    
    def al_mazo(sor)
      if(@barajada==false)
        @sorpresas << sor
      end
    end
    
    
    def siguiente()
      if(@barajada == false || @usadas == @sorpresas.size())
        @barajada = true
        @usadas=0
        @sorpresas.shuffle!
      end
      @usadas=@usadas+1
      @ultima_sorpresa=@sorpresas.at(0)
      @sorpresas.delete_at(0)
      @sorpresas << @ultima_sorpresa
      return @ultima_sorpresa
    end
    
    def inhabilitar_carta_especial(_sorpresa)
      for i in 0..(@sorpresas.size()-1)
        if(@sorpresas[i] == _sorpresa)
          @cartas_especiales << _sorpresa
          @sorpresas.delete_at(i)
          Diario.instance.ocurre_evento("Carta Sorpresa INHABILITADA")
        end
      end
    end
    
    def habilitar_carta_especial(_sorpresa)
      for i in 0..(@cartas_especiales.size()-1)
        if(@cartas_especiales[i] == _sorpresa)
          @sorpresas << _sorpresa
          @cartas_especiales.delete_at(i)
          Diario.instance.ocurre_evento("Carta Sorpresa HABILITADA")
        end
      end
    end
    
    def size()
      return @sorpresas.size()
    end
    
    def get_ultima_sorpresa()
      return @ultima_sorpresa
    end
    
    
  end
end

